### ISSUE TITLE
Segfault

### ISSUE THREAD

**Issue by GrbavaCigla** [fairyglade/ly#221]
I am on gentoo and when I try to run ly with agetty or openrc init script, segfault
agetty line I changed:
```
c2:2345:respawn:/sbin/agetty --skip-login --login-program /usr/bin/ly 38400 tty2 linux
```
OpenRC file I tried:
```sh
#!/sbin/openrc-run

description="start lu on a terminal line"
supervisor=supervise-daemon
port="8"
term_type="${term_type:-linux}"

command=/usr/bin/ly
agetty_options="-n -l /usr/bin/ly"
command_args_foreground="${agetty_options} ${port} ${term_type}"
pidfile="/run/${RC_SVCNAME}.pid"

depend() {
	after local
	keyword -prefix
}

start() {
	exec $command $command_args_foreground && chvt $port
	return $?
}

start_pre() {
	if [ -z "$port" ]; then
		eerror "${RC_SVCNAME} cannot be started directly. You must create"
		eerror "symbolic links to it for the ports you want to start"
		eerror "agetty on and add those to the appropriate runlevels."
		return 1
	else
		export EINFO_QUIET="${quiet:-yes}"
	fi
}

stop_pre()
{
	export EINFO_QUIET="${quiet:-yes}"
}
```

This is dmesg output:
```
ly[5717]: segfault at 10 ip 00005645a3fdcda2 sp 00007ffd17fffc50 error 4 in ly[5645a3fd6000+a000]
```

**Comment by kdkasad** [comment]
I am also getting regular segfaults. Mine happen after entering my password and pressing enter.

Here's the `dmesg` output:
```
[ 2473.459620] elogind[1379]: New session 46 of user kian.
[ 2473.468598] ly[6954]: segfault at 0 ip 00007f840be08a85 sp 00007ffcea512338 error 4 in libc-2.32.so[7f840bccc000+14d000]
[ 2473.468604] Code: 00 00 0f 1f 00 31 c0 c5 f8 77 c3 66 2e 0f 1f 84 00 00 00 00 00 f3 0f 1e fa 89 f9 48 89 fa c5 f9 ef c0 83 e1 3f 83 f9 20 77 2b <c5> fd 74 0f c5 fd d7 c1 85 c0 0f 85 eb 00 00 00 48 83 c7 20 83 e1
[ 2493.496188] elogind[1379]: Removed session 46.
```

**Comment by kdkasad** [comment]
Okay, after a bit of debugging, I found the problem. Calling `setenv(3)` with a value of NULL causes an error, which resulted in a segfault.

I created a fix for this:
```
The following changes since commit 77f6958241646e3f315f27bc38212c3c4e1e7a8d:

  update readme (2020-07-30 20:16:36 +0200)

are available in the Git repository at:

  https://git.kasad.com/ly.git for-upstream

for you to fetch changes up to 07d7009565417524c7b67743f1ebe8614ca96117:

  implement null/zero-length checks before calling setenv (2020-10-19 00:10:38 -0700)

----------------------------------------------------------------
Kian Kasad (1):
      implement null/zero-length checks before calling setenv

 src/login.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)
```

**Comment by GrbavaCigla** [comment]
Now it shows ly screen, but when I enter the password, it reloads. dmesg shows this:
```
[   23.139504] ly[2464]: segfault at 10 ip 000055aaf8f8de15 sp 00007ffcdee3c700 error 4 in ly[55aaf8f87000+a000]
[   35.228718] ly[2476]: segfault at 0 ip 000055e3db27e0a9 sp 00007fff5a63f5f0 error 4 in ly[55e3db27a000+a000]
[   41.144561] ly[2479]: segfault at 0 ip 000055e3db27e0a9 sp 00007fff5a63f5f0 error 4 in ly[55e3db27a000+a000]
```

**Comment by thinkoid** [comment]
FWIW, I investigated the same crash happening on my machine (artix), using `s6`. I am tinkering with making `ly` work under `s6` which does not work at the moment because whenever run by `s6` it crashes.

The `dmesg` output is:

```
[  889.091710] ly[4273]: segfault at 10 ip 000055c3055adf11 sp 00007ffe409ac110 error 4 in ly[55c3055a7000+b000]
[  889.091719] Code: 8c 00 00 83 f8 ff 74 0b 8b 05 83 8c 00 00 83 f8 ff 75 2c 83 7d fc ff 74 26 83 7d f8 ff 74 20 48 8b 05 cb 18 01 00 48 83 c0 10 <48> 8b 00 48 89 c6 48 8d 05 42 18 01 00 48 89 c7 e8 0f 1f 00 00 8b

```

Inspecting the coredump left behind I found out that the call to `tb_init` in `main` actually fails, leaving `termbox`'s `funcs` array uninitialized. When first used, the program crashes. The `tb_init` failure is in opening `/dev/tty` (actually in `tb_init_file` IIRC). 

It is worth adding a test for `tb_init` and only let it go forward if it succeeds.
